### ISSUE TITLE
"Re-run failed jobs" will not work with a parallel workflow

### ISSUE THREAD

**Issue by mellis481** [cypress-io/github-action#574]
After implementing a custom build ID to ensure I could re-run workflows which I've integrated with Cypress Dashboard and configured to run parallelly, I've run into an issue (oddly different than [this](https://github.com/cypress-io/github-action/issues/531) one).

Here is my job FWIW:
```
- id: cypress-mocked-api-tests
  uses: cypress-io/github-action@v2
  with:
    wait-on: 'https://localhost:9001/index.js'
    start: npm run start:${{ env.NODE_ENV }}
    config-file: cypress/config/${{ env.NODE_ENV }}.json
    config: video=true,videoUploadOnPasses=false
    spec: '**/*.spec.ts'
    install: false
    record: true
    parallel: true
    group: 'Mocked-API'
    ci-build-id: ${{ needs.prepare.outputs.uuid }}
```

This job will load balance all my spec files across five containers under a "Mocked-API" group.  This works great and I can re-run all jobs without issue.

On a recent run, one of the five containers failed because one test failed.  I thought I'd test how "Re-run failed jobs" worked on just the failed container job.  My hope/expectation was that it would be smart enough to know which spec files it ran when the entire workflow executed originally (which would have been six spec files which included 22 test) and run those.  Instead it ran zero spec files and completed successfully.  It seems like the matrix-level orchestration that is needed is not occurring when only a failed container job is re-run.  It looks like [someone else](https://github.community/t/how-to-disable-re-run-failed-jobs/238537) has run into this issue too and is trying to solve it by disabling the "Re-run failed jobs" option in Github (which doesn't seem possible).

This is a fairly big problem because it resulted in the group (which I've configured as a status check in my trunk branch protection rule) passing and the PR being able to be merged when it had never successfully run all tests.

**Comment by conversaShawn** [comment]
@mellis481 We recommend passing the GITHUB_TOKEN secret (created by the GH Action automatically) as an environment variable. This will allow correctly identifying every build and avoid confusion when re-running a build.

You can find an example here: https://github.com/cypress-io/github-action#record-test-results-on-cypress-dashboard

**Comment by mellis481** [comment]
@conversaShawn That did nothing. I did this:

- Added the GITHUB_TOKEN as an env variable to my `cypress-io/github-action@v2` job in my PR workflow.
- Added a failing test to my suite.
- Ran the workflow which is configured to run in parallel using 5 containers. The test failed on Machine 5.
- I executed "Re-run failed jobs".
- On the second workflow run, Machine 5 executed 0 tests and passed.

**Comment by mgambati** [comment]
Same case here.
Tests pass without execution after retrying failed jobs.

**Comment by modern-sapien** [comment]
related: [Github actions: "Re-run failed jobs" will run the entire test suite](https://github.com/cypress-io/github-action/issues/531)

**Comment by admah** [comment]
There were recently some changes in our services repo that may have taken care of this issue. Can someone retest with `10.7.0` or later and post results? Thanks!

**Comment by admah** [comment]
@mellis481 thanks for the screenshots and additional context. That's very helpful. I was able to get some more clarity on this from our Cloud team. 

Here is the current status:

- Before, there was an issue where all re-runs got a PASS, regardless of actual status. This issue has been fixed.
- Currently, if a re-run is initiated, all specs get run on the machines available. That is not optimal. The Cloud team is looking into the connection between GH Actions and Cypress in order to set up re-runs to be accurate and efficient.

**Comment by mellis481** [comment]
@admah I'm glad the update from your Cloud team matches my findings (in many less words :smile:).

**Comment by admah** [comment]
@mellis481 yes, I will be providing updates as they're available. 

I will be updating in #531 since this is actually a duplicate of that issue.

**Comment by admah** [comment]
Duplicate of #531

**Comment by jennifer-shehane** [comment]
@cgraham-rs If you're experiencing this behavior, it is because there is not a unique `ci-build-id` associated with the rerun. We try to interpret unique buildIds on our side, but if you're encountering this you can pass a unique buildId via the `ci-build-id` flag: https://docs.cypress.io/cloud/features/smart-orchestration/parallelization#Linking-CI-machines-for-parallelization-or-grouping

**Comment by cgraham-rs** [comment]
@jennifer-shehane as I mentioned in https://github.com/cypress-io/github-action/issues/431 the [Robust custom build id documentation](https://github.com/cypress-io/github-action/tree/master?tab=readme-ov-file#robust-custom-build-id) instructs us to manually craft a `ci-build-id`  as a separate job from the test job. Though it mentions `...If the user re-runs the workflow a new unique build id is generated...` which is true. But causes problems when a user chooses to `Re-run failed jobs` which would ONLY re-run the test job and not the setup job, thus the test job has the same `ci-build-id` from the previous run.

**Comment by jennifer-shehane** [comment]
@cgraham-rs This is an issue on our radar - that re-running failed jobs has a less than ideal experience in most CIs. We intend to look more into addressing this.

**Comment by cgraham-rs** [comment]
@jennifer-shehane A false pass is a really serious issue. AFAIK there's currently no open ticket tracking this specific problem. My suggestion is that this ticket be re-opened until such a time the specific scenario of Github `Re-run failed jobs` cannot run 0 tests and throw a false pass.

**Comment by danjohansenconsulting** [comment]
@cgraham-rs we have an initiative in Cypress Cloud that is a precursor for support of Re-run failed jobs.  That initial work is scheduled for Q1.  Once we have a solution that launches we will announce the release of that in Cloud.

**Comment by mellis481** [comment]
>  This is an issue on our radar - that re-running failed jobs has a less than ideal experience in most CIs. We intend to look more into addressing this.

@jennifer-shehane To @cgraham-rs's question, is there a ticket that can be tracked? And could you re-open this ticket please? It should never have been marked as a dupe of 531. 

> we have an initiative in Cypress Cloud that is a precursor for support of Re-run failed jobs. That initial work is scheduled for Q1. Once we have a solution that launches we will announce the release of that in Cloud.

@danjohansenconsulting Can you provide an update on this please?

**Comment by MikeMcC399** [comment]
@mellis481 

There is an update to this theme in Discord https://discord.com/channels/755913899261296641/1382740640630313010/1392223502140051466 dated July 8, 2025

<img width="778" height="103" alt="Image" src="https://github.com/user-attachments/assets/47a2796b-e54b-425e-9480-702fd7585707" />

**Comment by MikeMcC399** [comment]
- FWIW https://github.com/cypress-io/github-action/issues/1290 is the related open issue in this repo, however it's really only a placeholder as the main code changes won't be in this repo, so the issue can only be manually maintained, and that will need participation from the Cypress Cloud team
